crates/aiken-lang/src/tipo/expr.rs

```diff
@@ -975,15 +965,21 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
                 });
             }
 
-            // Ensure the pattern matches the type of the value
-            PatternTyper::new(self.environment, &self.hydrator).unify(
-                untyped_pattern.clone(),
-                value_typ.clone(),
-                None,
-                kind.is_let(),
-            )?
+            None
         };
 
+        // Ensure the pattern matches the type of the value
+        let pattern = PatternTyper::new(self.environment, &self.hydrator).unify(
+            untyped_pattern.clone(),
+            value_typ.clone(),
+            ann_typ,
+            kind.is_let(),
+        )?;
+
+        if kind.is_expect() && value_typ.contains_opaque() {
```

Comment:
This should only throw if value_typ is Data, not if value type has already been casted. expecting on a non-Data value is always fine.